doc/esp8266wifi/generic-class.rst

```diff
@@ -42,6 +42,33 @@ persistent
 
     WiFi.persistent(persistent)
 
+Starting from version 3 of this core, **persistence is disabled by default
+and WiFi does not anymore automatically fires up at boot** (see PR `#7902 <https://github.com/esp8266/Arduino/pull/7902>`__).
+
+Previously, SDK was automatically starting WiFi at boot.  This was probably
+intended for the Espressif AT FW which is interactive and preserves WiFi
+state accross reboots.  This behavior is generally irrelevant with the
+Arduino API because sketches start with ``WiFi.begin()`` or
+``WiFi.softAP()``.
+
+This change is harmless with standard sketches: Calls to ``WiFi.mode()`` do
+enable radio as usual.  It also smooths current spikes at boot and decreases
+DHCP stress.
+
+Legacy behavior can be restored by calling ``enableWiFiAtBoot()`` from
```

Comment:
To restore legacy behavior, in addition to calling enableWiFiAtBoot() the user has to call WiFi.persitent(true). This should be used with care, because use could lead to early wear of the wifi config flash sector and hence device failure.

Reply:
See my comment about, `WiFi.persistent(true)` is only needed to change the flash credentials cache, not in a sketch that just wants to use whatever was persisted at any time prior to it.

Reply:
> the user has to call WiFi.persitent(true).

True !
I will add it in `enableWiFiAtBootTime()`

> could lead to early wear of the wifi config flash sector

Given the fact that when persistence is enabled, [user config is compared to stored config](https://github.com/esp8266/Arduino/blob/1b922edad127fa515dc7a00128f9cabfe0b1a67b/libraries/ESP8266WiFi/src/ESP8266WiFiSTA.cpp#L171), I think that's OK.
What do you both think ?

Reply:
I always thought this mode was very important !!!

Reply:
When the wifi is active there is a consumption of energy that it would be good to keep in case of not use, especially if you use batteries for their power supply. In case of sending data, as sometimes I do, it is good to write to the internal memory and then make only one connection per day and send all the data recorded in the preset intervals.

Reply:
@Bighoneypot The two WiFi / low power examples should work as before:
https://github.com/esp8266/Arduino/tree/master/libraries/esp8266/examples/LowPowerDemo
https://github.com/esp8266/Arduino/blob/master/libraries/ESP8266WiFi/examples/WiFiShutdown/WiFiShutdown.ino

Reply:
Maybe also good to test with a number of different ESP modules (ESP12E/F/S for example) to see if this also has an effect on deep sleep power consumption.
There have been a number of reports on deep sleep power consumption not being close to 60 uA on some nodes, with some reporting success with the most unclear assembly blobs possible and others claiming you need to do a short deep sleep first and then again call deep sleep to get true low power deep sleep and others reporting the WiFi will not start after deep sleep until you go into deep sleep again.

So I would not be surprised to see such topics pop up again if this is changed. (or this may be the final fix for it.... who knows ;) )

Reply:
If you need deep sleep and low power, you need to set up a series of special things, such as early wifi init, persistent wifi creds, static IP, etc, and then your app will usually have something in setup() and an empty loop. Then you need to check power and optimize more. That's an entirely different and very specific use case compared to the basic "start wifi, connect to some router, wiggle some pins" sketch that newer users usually start with. 
In other words, deep sleep needs a special kind of sketch unlike the typical Arduino style setup - loop case addressed by this PR.

Reply:
> @Bighoneypot The two WiFi / low power examples should work as before:
> https://github.com/esp8266/Arduino/tree/master/libraries/esp8266/examples/LowPowerDemo
> https://github.com/esp8266/Arduino/blob/master/libraries/ESP8266WiFi/examples/WiFiShutdown/WiFiShutdown.ino

Thanks i watch this.

Reply:
> Maybe also good to test with a number of different ESP modules (ESP12E/F/S for example) to see if this also has an effect on deep sleep power consumption.
> There have been a number of reports on deep sleep power consumption not being close to 60 uA on some nodes, with some reporting success with the most unclear assembly blobs possible and others claiming you need to do a short deep sleep first and then again call deep sleep to get true low power deep sleep and others reporting the WiFi will not start after deep sleep until you go into deep sleep again.
> 
> So I would not be surprised to see such topics pop up again if this is changed. (or this may be the final fix for it.... who knows ;) )

I didn't measure any difference in the Deep Sleep current.  I looked at startup current and different types of Sleep.  The only expected difference with other boards/modules would be due to extra chips on the boards (LDO, USB chip, battery charger, etc.), You can get several uA of difference if your GPIO outputs go the opposite direction from the pullup/pulldown resistors on those pins, ~2uA per pin in that case or 6-8uA higher total, but you pretty much have to have the pullup/pulldown resistors either on the module or in your circuit or it won't boot to Flash.

The people that needed to reset across Deep Sleep a second time to re-enable WiFi after ESP.deepSleep(uS, WAKE_RF_DISABLED) have a fix [here](https://github.com/esp8266/Arduino/issues/3072#issuecomment-741524721).  The only wart is that it ALWAYS does a full RF_CAL when you wake the modem that way.  I guess it blanks out the stored values for the modem parameters when you do WAKE_RF_DISABLED, so it has to rebuild everything from scratch.

As far as variable current during Deep Sleep, I've never seen it mentioned.  Most likely they weren't in Deep Sleep, for whatever reason.  The SDK stops WiFi and all timers before entering Sleep, and once it's asleep the CPU PLL screeches to a complete halt and the power drops to around 20uA, which is just the RTC timer running.  I get closer to 16uA since I'm frequently running at 3.0V instead of 3.3V.  If someone has an MCVE that demonstrates variable Deep Sleep currents I can figure out what's going on.

I _have_ seen numerous threads about Deep Sleep Zombie mode where it never wakes properly, but the jury is still out on the actual cause(s).  I vote for cheap Chinese Flash that doesn't always come up in full drive mode after the Flash has been slept prior to Deep Sleep, but Erriez thinks it's the NodeMCU reset circuit interfering.  We could both be right: two different causes or more.  Until I get one here I can dissect, my tools are useless at troubleshooting across the Internet.